### DISCUSSION TITLE
Link error when building zoul tests

### DISCUSSION THREAD

**Discussion by mizapf** [contiki-ng/contiki-ng#1567] [Q&A]
I've got a bundle of Zoul Re-Motes RevB, so I checked out the contiki-ng project and tried to build the test programs in examples/platform-specific/zoul.

However, I end up with this (the same happens with contiki-os):

$ make TARGET=zoul 
which: no rlwrap in (/home/michael/bin:/usr/local/bin:/usr/bin:/bin)
  MKDIR     build/zoul/remote-revb/obj
  CC        ../../../arch/platform/zoul/dev/tsl256x.c
  CC        ../../../arch/platform/zoul/dev/sht25.c
  CC        ../../../arch/platform/zoul/dev/bmpx8x.c
  CC        ../../../arch/platform/zoul/dev/motion-sensor.c
  CC        ../../../arch/platform/zoul/dev/adc-sensors.c
  CC        ../../../arch/platform/zoul/dev/weather-meter.c
[ ... more lines ... ]
  CC        ../../../os/net/routing/rpl-lite/rpl-of0.c
  CC        ../../../os/net/routing/rpl-lite/rpl-timers.c
  CC        ../../../arch/cpu/cc2538/cc2538.lds
  CC        ../../../arch/cpu/cc2538/./startup-gcc.c
  CC        test-tsl256x.c
  LD        build/zoul/remote-revb/test-tsl256x.elf
/usr/lib64/gcc/arm-none-eabi/10/ld: build/zoul/remote-revb/obj/bme280-sensor.o:(.bss.bme280_mea+0x0): multiple definition of `bme280_mea'; build/zoul/remote-revb/obj/bme280.o:(.bss.bme280_mea+0x0): first defined here
collect2: error: ld returned 1 exit status
make: *** [../../../arch/cpu/arm/cortex-m/Makefile.cortex-m:26: build/zoul/remote-revb/test-tsl256x.elf] Error 1
rm test-tsl256x.o build/zoul/remote-revb/obj/startup-gcc.o

Any ideas what I may be doing wrong here?

**Comment by g-oikonomou** [comment]
I just tried locally and it works, so I am suspecting a toolchain version thing - you are not necessarily doing anything wrong. Perhaps you have a newer toolchain version that complains about something that older ones don't?

I tried with the version below on macos

```
$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 7-2017-q4-major) 7.2.1 20170904 (release) [ARM/embedded-7-branch revision 255204]
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by mizapf** [comment]
I installed the cross compiling toolchain on my Tumbleweed Linux (openSUSE).

$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (SUSE Linux) 10.2.1 20201202 [revision e563687cf9d3d1278f45aaebd03e0f66531076c9]
Copyright (C) 2020 Free Software Foundation, Inc.
...

Still weird that such a thing as duplicate symbols could change over toolchain versions.

**Reply by wdargie** [reply]
I am facing the same problem with  TARGET=zoul. How did you solve the problem? Can you share please?

**Reply by pjonsson** [reply]
@wdargie Are you using Contiki-NG v4.7 or later?

**Reply by wdargie** [reply]
Version 4.8

**Reply by mizapf** [reply]
I'm not sure whether I had the same issues as in that post of mine from 2021; in the meantime I succeeded installing contiki-ng in an openSUSE Linux desktop distribution (without Docker stuff). 

I downloaded the Arm GNU toolchain from developer.arm.com (gcc-arm-11.2-2022.02-x86_64-arm-none-eabi.tar.xz), unpacked it in a folder in my home directory, and added its bin folder to the head of my PATH. I noticed that there may be issues with such cross-compiling tools as provided by SUSE.

From contiki-ng I use the develop branch; I just pulled the recent commit and tried to compile hello-world; all OK and running.

**Reply by wdargie** [reply]
Try an older version of arm-none-eabi-gcc (say, v. 19). For the zoul TARGET, that solved the problem for me.

**Reply by mizapf** [reply]
I tried some versions of the cross-compiler from ARM. The command line is always
$ **make hello-world TARGET=zoul BOARD=remote-revb**

**11.2**: builds correctly

**11.3**:   [...] LD        build/zoul/remote-revb/hello-world.elf
/home/michael/prg/arm-gnu-toolchain-11.3.rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld:build/zoul/remote-revb/obj/cc2538.ld:1503: warning: memory region `FLASH_FW=' not declared
/home/michael/prg/arm-gnu-toolchain-11.3.rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/11.3.1/../../../../arm-none-eabi/bin/ld:build/zoul/remote-revb/obj/cc2538.ld:1503: syntax error
collect2: error: ld returned 1 exit status

**12.2**:   [...] CC        ../../arch/cpu/cc2538/dev/aes.c
In file included from ../../arch/cpu/cc2538/dev/aes.c:44:
../../arch/cpu/cc2538/dev/aes.c: In function 'aes_load_keys':
../../arch/cpu/cc2538/./dev/rom-util.h:118:16: error: array subscript 0 is outside array bounds of 'struct rom_util_api[0]' [-Werror=array-bounds]
  118 |   (ROM_UTIL_API->memcpy((dest), (src), (n)))
      |                ^~
../../arch/cpu/cc2538/dev/aes.c:68:7: note: in expansion of macro 'rom_util_memcpy'
   68 |       rom_util_memcpy(&aligned_keys[i << 2], &((const uint64_t *)keys)[i * 3],
      |       ^~~~~~~~~~~~~~~
../../arch/cpu/cc2538/./dev/rom-util.h:118:16: error: array subscript 0 is outside array bounds of 'struct rom_util_api[0]' [-Werror=array-bounds]
  118 |   (ROM_UTIL_API->memcpy((dest), (src), (n)))
      |                ^~
../../arch/cpu/cc2538/dev/aes.c:81:5: note: in expansion of macro 'rom_util_memcpy'
   81 |     rom_util_memcpy(aligned_keys, keys, count << 4);
      |     ^~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

**Reply by pjonsson** [reply]
Thank you @mizapf, there aren't many remote-rev boards used in the build tests for some reason, we'll take a look at fixing those issues.

**Reply by pjonsson** [reply]
There is still some work required for building Contiki-NG with GCC 12, but the issues shown here should have been fixed in the develop branch for ARM-based targets.

**Comment by pjonsson** [comment]
I think @Yagoor fixed this with commit 8cf35b00973876b8127ee4875bb9d58dab526f97.

**Comment by wdargie** [comment]
Yes.


<http://wwwpub.zih.tu-dresden.de/~dargie/>

________________________________
Von: Peter A. Jonsson ***@***.***>
Gesendet: Mittwoch, 17. Mai 2023 13:22:47
An: contiki-ng/contiki-ng
Cc: Dargie, Waltenegus; Mention
Betreff: Re: [contiki-ng/contiki-ng] Link error when building zoul tests (Discussion #1567)


@wdargie<https://github.com/wdargie> Are you using Contiki-NG v4.7 or later?

**Reply by pjonsson** [reply]
The `examples/zoul` directory is built by the CI in `tests/02-compile-arm-ports`, so the missing symbols does not occur on the develop branch with the toolchain used by the Docker image. What version of the compiler are you running, and does your tree have the changes from commit [`8cf35b0`](https://github.com/contiki-ng/contiki-ng/commit/8cf35b00973876b8127ee4875bb9d58dab526f97)?

**Comment by wdargie** [comment]
It is version 12.2.1 (20230214). It seems that I get different errors whenever I change the  ARM''s version...





Dr.-Ing. habil. Waltenegus Dargie
Visiting Professor, University of the Witwatersrand, Johannesburg

Courtesy Assistant Professor, Knight Foundation School of Computer and Information Sciences, Florida International University
Head of Energy Lab

Erasmus Coordinator

Faculty of Computer Science
TU Dresden
01062 Dresden
GERMANY
Tel.: +49 (0)351 463 38352
http://wwwpub.zih.tu-dresden.de/~dargie/


<http://wwwpub.zih.tu-dresden.de/~dargie/>

________________________________
Von: Peter A. Jonsson ***@***.***>
Gesendet: Mittwoch, 17. Mai 2023 14:39:10
An: contiki-ng/contiki-ng
Cc: Dargie, Waltenegus; Mention
Betreff: Re: [contiki-ng/contiki-ng] Link error when building zoul tests (Discussion #1567)


The examples/zoul directory is built by the CI in tests/02-compile-arm-ports, so the missing symbols does not occur on the develop branch with the toolchain used by the Docker image. What version of the compiler are you running, and does your tree have the changes from commit 8cf35b0<https://github.com/contiki-ng/contiki-ng/commit/8cf35b00973876b8127ee4875bb9d58dab526f97>?

**Reply by pjonsson** [reply]
The reason the docker image runs 10.3 was because there was some issue with newer compiler versions from ARM not running on older machines. Contiki-NG should still be buildable with more recent compiler versions though, could you paste the build output for the error you are getting?

**Reply by pjonsson** [reply]
Contiki-NG should not force you to run a compiler from 2019 if you are targeting ARM, here's the GCC version that is used by the CI tests (in the docker image) without linking errors:
```
$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (GNU Arm Embedded Toolchain 10.3-2021.10) 10.3.1 20210824 (release)
...
```

**Reply by wdargie** [reply]
Again, I tried with Version 10.3-2021.10 but had a linking problem on Ubuntu. When I replaced it with Version 9-2020-q2-update, the irritation was gone.
